batcher/batcher/cardano_chain_operations_strategy.go

```diff
@@ -0,0 +1,252 @@
+) (*cardano.TxOutputs, uint64, error) {
+	receiversMap := map[string]uint64{}
+
+	for _, transaction := range txs {
+		for _, receiver := range transaction.Receivers {
+			receiversMap[receiver.DestinationAddress] += receiver.Amount.Uint64()
+		}
+	}
+
+	result := cardano.TxOutputs{
+		Outputs: make([]cardanowallet.TxOutput, 0, len(receiversMap)),
+		Sum:     map[string]uint64{},
+	}
+
+	for addr, amount := range receiversMap {
+		if amount == 0 {
+			logger.Warn("skipped output with zero amount", "addr", addr)
+
+			continue
+		} else if !cardano.IsValidOutputAddress(addr, cardanoConfig.NetworkID) {
+			// apex-361 fix
+			logger.Warn("skipped output because it is invalid", "addr", addr)
+
+			continue
+		}
+
+		result.Outputs = append(result.Outputs, cardanowallet.TxOutput{
+			Addr:   addr,
+			Amount: amount,
+		})
+		result.Sum[cardanowallet.AdaTokenName] += amount
+	}
+
+	// sort outputs because all batchers should have same order of outputs
+	sort.Slice(result.Outputs, func(i, j int) bool {
+		return result.Outputs[i].Addr < result.Outputs[j].Addr
+	})
+
+	return &result, 0, nil
+}
+
+func (s *CardanoChainOperationReactorStrategy) GetNeededUtxos(
+	inputUTXOs []*indexer.TxInputOutput,
+	desiredAmount map[string]uint64,
+	minUtxoAmount uint64,
+	utxoCount int,
+	maxUtxoCount int,
+	_ uint64,
+	takeAtLeastUtxoCount int,
+) (chosenUTXOs []*indexer.TxInputOutput, err error) {
+	inputUTXOs = filterOutTokenUtxos(inputUTXOs)
+	// if we have change then it must be greater than this amount
+	desiredAmount[cardanowallet.AdaTokenName] += minUtxoAmount
+
+	inUtxos := mapUtxos(inputUTXOs)
+
+	outputUTXOs, err := bridgingtx.GetUTXOsForAmounts(inUtxos, desiredAmount, maxUtxoCount, takeAtLeastUtxoCount)
+	if err != nil {
+		return nil, err
+	}
+
+	usedUtxoMap := map[string]bool{}
+	for _, utxo := range outputUTXOs.Inputs {
+		usedUtxoMap[utxo.String()] = true
+	}
+
+	chosenCount := 0
+
+	for _, utxo := range inputUTXOs {
+		if !usedUtxoMap[fmt.Sprintf("%s#%d", utxo.Input.Hash, utxo.Input.Index)] {
+			continue
+		}
+
+		chosenUTXOs = append(chosenUTXOs, utxo)
+		chosenCount++
+	}
+
+	return chosenUTXOs, nil
+}
+
+type CardanoChainOperationSkylineStrategy struct {
+}
+
+func (s *CardanoChainOperationSkylineStrategy) GetOutputs(
+	txs []eth.ConfirmedTransaction, cardanoConfig *cardano.CardanoChainConfig,
+	destChainID string, logger hclog.Logger,
+) (*cardano.TxOutputs, uint64, error) {
+	receiversMap := map[string]cardanowallet.TxOutput{}
+
+	var tokenHoldingOutputs uint64 = 0
+
+	for _, transaction := range txs {
+		for _, receiver := range transaction.Receivers {
+			data := receiversMap[receiver.DestinationAddress]
+			data.Amount += receiver.Amount.Uint64()
+
+			if receiver.AmountWrapped != nil && receiver.AmountWrapped.Sign() > 0 {
+				if len(data.Tokens) == 0 {
+					tconf := getConfigTokenExchange(destChainID, true, cardanoConfig.Destinations)
+					token, err := cardanowallet.NewTokenAmountWithFullName(tconf.DstTokenName, 0, true)
```

Comment:
neatpick (formatting): `token, err := ...` and error check should not be separated